geps/gep-1762.md

```diff
@@ -0,0 +1,281 @@
+
+## Goals
+
+* Provide prescriptive guidance for how in-cluster implementations should behave.
+* Provide requirements (tested by conformance) for how in-cluster implementations should behave.
+
+Note that some changes will be suggestions, while others will be requirements.
+
+## Non-Goals
+
+* Provide guidance to how out-of-cluster implementations should behave. Rather, this document aims to bring consistency between these types.
+
+## Terminology
+
+This document uses a few terms throughout. To ensure consistency, they are defined below:
+
+* In-cluster deployment: refers to an implementation that actuates a `Gateway` by running a data plane in the cluster.
+  This is *often*, but not necessarily, by deploying a `Deployment`/`DaemonSet` and `Service`.
+* Automated deployment: refers to an implementation that automatically deploys the data plane based on a `Gateway`.
+  That is, the user simply creates a `Gateway` resource and the rest is handled behind the scenes by the implementation.
+* Manual deployment: refers to an implementation that does NOT automatically deploy the data plane based on a `Gateway`.
+  This may require, for example, a user to manually create a `Deployment` and `Service` for the data plane, then link this to a `Gateway`.
+
+## Design
+
+This GEP does not introduce new API fields, but rather standardizes how implementations should behave when implementing the existing API.
+
+### Automated Deployments
+
+A simple `Gateway`, as is configured below is assumed to be an automated deployment:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: my-gateway
+spec:
+  gatewayClassName: example
+  listeners:
+  - name: default
+    port: 80
+    protocol: HTTP
+```
+
+With this configuration, an implementation:
+* MUST mark the Gateway as `Programmed` and provide an address in `Status.Addresses` where the Gateway can be reached on each configured port.
+* MUST label all generated resources (Service, Deployment, etc) with `gateway.networking.k8s.io/metadata.name: my-gateway` (where `my-gateway` is the name of the Gateway resource).
+* MUST provision generated resources in the same namespace as the Gateway if they are namespace scoped resources.
+  * Cluster scoped resources are not recommended.
+* MUST name all generated resources `my-gateway-example` (`<NAME>-<GATEWAY CLASS>`).
+  This is not simply `NAME` to reduce the chance of conflicts with existing resources.
+
+### Manual Deployments
+
+When a manual deployment is used, a user must explicitly link their Gateway to existing infrastructure. This is done by specifying `spec.infrastructure.attachTo`.
+
+For the common case that the "existing infrastructure" is a `Service`:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: my-gateway
+spec:
+  infrastructure:
+    attachTo:
+      kind: Service
+      name: existing-service
```

Comment:
This is currently a single object, but adequately referencing resources may require multiple.

In Kong's case, we validated that the resources associated with our manually-deployed Gateway could satisfy the Gateway (i.e. that the indicated Listeners were actually configured). We needed to check both the Service and the Deployment, since the former handles the LB provisioning and the latter handled application-level port binds.

Reply:
Yeah I guess we could extend it. Although I am curious why you need to reference the Deployment for ports - isn't that handled by TargetPort?

Reply:
It's for the case where you've done something like `targetPort: 80` in the Kubernetes configuration, but don't have something equivalent to a `listen 80` in an nginx.conf. We want to ensure there's a matching configuration for the target ports.

Reply:
Ah in our case the control plane configures the proxy config, so its impossible to end up in that state. Thanks for the info